### PR TITLE
Don't log first_difference & different_keys fields at all when only doing a quick comparison

### PR DESCRIPTION
**Pull request by aldavidson**
[Trello card](https://trello.com/c/sv4yzHj4/865-fix-problem-of-full-comparison-log-lines-from-the-content-store-proxy-not-appearing-in-logit)
After deploying #43 at around 17:45 on Friday 6th, we saw a steady percentage of fully-compared responses for a while, but  zero since `8 Oct, 2023 @ 00:15:02` on integration, and since `7 Oct, 2023 @ 00:15:03` on staging.

On investigation, it looks like fully-compared lines do get generated, but they don't make it into Logit. It seems Logit is refusing to parse them - probably due to a conflicting format with the (more-prevalent) "N/A" string-valued quick comparison lines. 

The cut off at 00:15 on successive nights suggests that either Logit updated their code, or (more likely) that some overnight job is being run to update index patterns and infer field formats from log entries, and the first entry it came across was a quick comparison with string-values. All subsequent lines with structured values (i.e. full comparison) are then rejected as invalid. 

So this PR:

* Removes the `first_difference` & `different_keys` fields entirely when only doing a quick comparison
* Also logs the randomly-generated number & the configured sample percentage
* Refactors the calculation out of the app and into the `ResponseComparator` object to make this simpler
* Refactors the `ResponseComparator` to use an instance with attributes, rather than stateless class methods. This makes for simpler code, as many of the method arguments can now be removed.

I've deployed this branch to integration, and we can see that the full-comparison lines are now making it into Logit again:
 
Screenshot from 2023-10-10 08-41-46

This application is owned by the publishing platform team. Please let us know in #govuk-publishing-platform when you raise any PRs.   

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

File: lib/response_comparator.rb [hunk 1]
```diff
@@ -25,6 +25,13 @@ def self.quick_comparison(primary_response, secondary_response)
     }
   end
 
+  def self.full_comparison?(comparison, full_pct)
+    r = rand(99)
```
Comment (jkempster34) [inline lib/response_comparator.rb]: Apparently `rand` only returns random numbers between 0 and the argument (not including the argument itself), so perhaps this should be `rand(100)` to capture 0-99

File: lib/response_comparator.rb [hunk 2]
```diff
@@ -25,6 +25,13 @@ def self.quick_comparison(primary_response, secondary_response)
...
+    full = (full_pct == 100 || r < full_pct)
```
Comment (jkempster34) [inline lib/response_comparator.rb:36]: If the above comment is taken into account, could this `full_pct == 100` be removed so the entire method would be 🤔 

```ruby
def self.full_comparison?(comparison, full_pct)
    r = rand(100) # between 0 - 99
    comparison.merge!(sample_percent: full_pct, r:)
    full
  end
```

**Review by jkempster34 — APPROVED** [review]
LGTM ✅

File: lib/response_comparator.rb (hunk 2 above)
Comment (aldavidson) [reply lib/response_comparator.rb]: good call, simplified as per your suggestion - thanks!

**Pull request merged**